### PR TITLE
Add parameter validation for AES module

### PR DESCRIPTION
**Pull request by mpg**
## Description

Finish adding parameter validation to the AES module (some functions where done in #1643 as a demo).

This PR is based on #1643, please only review new commits, starting from "Add parameter validation XTS setkey functions".

## Status
**READY**

## Requires Backporting
NO  

## Migrations
None in the default config.

### CODE REVIEW

File: include/mbedtls/aes.h [hunk 1]
```diff
@@ -246,10 +246,13 @@ int mbedtls_aes_xts_setkey_dec( mbedtls_aes_xts_context *ctx,
  *                 call to this API with the same context.
  *
  * \param ctx      The AES context to use for encryption or decryption.
+ *                 It must be initialized and bound to a key.
  * \param mode     The AES operation: #MBEDTLS_AES_ENCRYPT or
  *                 #MBEDTLS_AES_DECRYPT.
- * \param input    The 16-Byte buffer holding the input data.
- * \param output   The 16-Byte buffer holding the output data.
+ * \param input    The buffer holding the input data.
+ *                 It must be readable and at least 16 Bytes long.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Use \c.

File: tests/suites/test_suite_aes.function [hunk 2]
```diff
@@ -406,6 +408,20 @@ void aes_invalid_param( )
...
                             mbedtls_aes_xts_setkey_dec( &xts_ctx, NULL, 128 ) );
 #endif
+
+
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_AES_BAD_INPUT_DATA,
+                            mbedtls_aes_crypt_ecb( NULL,
+                                                   MBEDTLS_AES_ENCRYPT, in, out ) );
+    TEST_INVALID_PARAM_RET( MBEDTLS_ERR_AES_BAD_INPUT_DATA,
+                            mbedtls_aes_crypt_ecb( &aes_ctx,
+                                                   42, in, out ) );
```
Comment (hanno-arm) [inline tests/suites/test_suite_aes.function]: Minor: As suggest by @k-stachowiak in another PR, use named constants for valid and invalid modes. I.e. `int valid_mode = MBEDTLS_AES_ENCRYPT`, `int invalid_mode = 42`. Not a blocker.

File: include/mbedtls/aes.h [hunk 3]
```diff
@@ -287,13 +287,17 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
  *
  *
  * \param ctx      The AES context to use for encryption or decryption.
+ *                 It must be initialized and bound to a key.
  * \param mode     The AES operation: #MBEDTLS_AES_ENCRYPT or
  *                 #MBEDTLS_AES_DECRYPT.
  * \param length   The length of the input data in Bytes. This must be a
  *                 multiple of the block size (16 Bytes).
  * \param iv       Initialization vector (updated after use).
+ *                 It must be a readable and writeable buffer of 16 Bytes.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Use \c.

File: include/mbedtls/aes.h [hunk 4]
```diff
@@ -287,13 +287,17 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
...
  * \param input    The buffer holding the input data.
+ *                 It must be readable and of size \p length.
  * \param output   The buffer holding the output data.
+ *                 It must be writeable and of size \p length.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Add `Bytes`.

File: library/aes.c [hunk 5]
```diff
@@ -1143,6 +1182,12 @@ int mbedtls_aes_crypt_xts( mbedtls_aes_xts_context *ctx,
     unsigned char prev_tweak[16];
     unsigned char tmp[16];
 
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply library/aes.c]: I think we should add this check here, too.

File: include/mbedtls/aes.h [hunk 6]
```diff
@@ -380,13 +380,18 @@ int mbedtls_aes_crypt_xts( mbedtls_aes_xts_context *ctx,
...
+ *                 It must be initialized and bound to a key.
  * \param mode     The AES operation: #MBEDTLS_AES_ENCRYPT or
  *                 #MBEDTLS_AES_DECRYPT.
  * \param length   The length of the input data.
  * \param iv_off   The offset in IV (updated after use).
+ *                 It must point to a valid \c size_t.
  * \param iv       The initialization vector (updated after use).
+ *                 It must be a readable and writeable buffer of 16 Bytes.
  * \param input    The buffer holding the input data.
+ *                 It must be readable and of size \p length.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Add "Bytes".

File: include/mbedtls/aes.h [hunk 7]
```diff
@@ -380,13 +380,18 @@ int mbedtls_aes_crypt_xts( mbedtls_aes_xts_context *ctx,
  *
  *
  * \param ctx      The AES context to use for encryption or decryption.
+ *                 It must be initialized and bound to a key.
  * \param mode     The AES operation: #MBEDTLS_AES_ENCRYPT or
  *                 #MBEDTLS_AES_DECRYPT.
  * \param length   The length of the input data.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Add "in Bytes".

File: library/aes.c [hunk 8]
```diff
@@ -1382,7 +1382,15 @@ int mbedtls_aes_crypt_ofb( mbedtls_aes_context *ctx,
...
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( iv_off != NULL );
+    AES_VALIDATE_RET( iv != NULL );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+
+    n = *iv_off;
 
     while( length-- )
```
Comment (hanno-arm) [inline library/aes.c]: Pre-existing: We're missing an unconditional validation of `*iv_off` here. A bad value leads to an unlimited buffer overread.

File: include/mbedtls/aes.h [hunk 9]
```diff
@@ -561,15 +561,21 @@ int mbedtls_aes_crypt_ofb( mbedtls_aes_context *ctx,
...
  *                         resuming within the current cipher stream. The
  *                         offset pointer should be 0 at the start of a stream.
+ *                         It must point to a valid \c size_t.
  * \param nonce_counter    The 128-bit nonce and counter.
+ *                         It must be a readable-writeable buffer of 16 Bytes.
  * \param stream_block     The saved stream block for resuming. This is
  *                         overwritten by the function.
+ *                         It must be a readable-writeable buffer of 16 Bytes.
  * \param input            The buffer holding the input data.
+ *                         It must be readable and of size \p length.
```
Comment (hanno-arm) [inline include/mbedtls/aes.h]: Minor: Add "Bytes"

File: include/mbedtls/padlock.h [hunk 10]
```diff
@@ -3,6 +3,9 @@
  *
  * \brief VIA PadLock ACE for HW encryption/decryption supported by some
  *        processors
+ *
+ * \warning  These functions are only for internal use by other library
+ * functions; you must not call them directly.
```
Comment (hanno-arm) [inline include/mbedtls/padlock.h]: Minor: Indentation.

File: library/aes.c [hunk 11]
```diff
@@ -981,6 +1006,12 @@ int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
                     const unsigned char input[16],
```
Comment (sbutcher-arm) [inline library/aes.c]: Indentation is out. (Not a blocker).

File: library/aes.c [hunk 12]
```diff
@@ -1289,6 +1344,12 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
     unsigned char c;
```
Comment (sbutcher-arm) [inline library/aes.c]: Indentation is out from lines 1338 onwards. (Not a blocker, and pre-existing).

File: library/aes.c [hunk 13]
```diff
@@ -1392,6 +1394,9 @@ int mbedtls_aes_crypt_ofb( mbedtls_aes_context *ctx,
 
     n = *iv_off;
 
+    if( n > 16 )
```
Comment (hanno-arm) [inline library/aes.c]: I think `n == 16` is already invalid.

**Comment by mpg** [comment]
@hanno-arm Thanks for catching the missing `iv_off` check! I've added that, and addressed the rest of your feedback as well.

@sbutcher-arm Thanks for your feedback. I've addressed most of it, except:
- I'd rather not touch the error codes at this point: unlike the other modules, here we already had the 3 of them so it's different from adding a new one. IMO there are at least two differences: we're not adding any new code, so not making things worse in terms of error code space usage. And also, changing the error code would mean changing existing behaviour that people might be relying upon.
- I agree about allowing NULL if length is 0, but as it's extra work for adding the tests, I'd rather leave it untouched for now and introduce it as a further enhancement later.

@hanno-arm @sbutcher-arm Please review again.

**Comment by mpg** [comment]
Just pushed a new commit to fix unused parameter warnings in test-ref-config. @hanno-arm @sbutcher-arm Please review again.

File: tests/suites/test_suite_aes.function [hunk 14]
```diff
@@ -563,13 +563,21 @@ void aes_check_params( )
...
+#endif
 #if defined(MBEDTLS_CIPHER_MODE_XTS)
     mbedtls_aes_xts_context xts_ctx;
 #endif
-    const unsigned char in[16] = { 0 };
-    unsigned char out[16];
+#if defined(MBEDTLS_CIPHER_MODE_CFB) || \
+    defined(MBEDTLS_CIPHER_MODE_OFB)
     size_t size;
+#endif
```
Comment (hanno-arm) [inline tests/suites/test_suite_aes.function]: I think in this instance the code would be easier to read and maintain if we unconditionally included `((void) xxx)`, but that's of course only a comment and not a change request.

**Comment by sbutcher-arm** [comment]
Jenkins CI fails only with the BSD timing test, which is as good as a pass. Promoting to 'ready for merge'.

**Comment by sbutcher-arm** [comment]
About deprecating `MBEDTLS_ERR_AES_INVALID_KEY_LENGTH`:

> I'd rather not touch the error codes at this point: unlike the other modules, here we already had the 3 of them so it's different from adding a new one. IMO there are at least two differences: we're not adding any new code, so not making things worse in terms of error code space usage. And also, changing the error code would mean changing existing behaviour that people might be relying upon.

We only have the three error codes, because I created one in the proof of concept PR!

I think we're creating an inconsistency across the interfaces here, as Blowfish, Camellia, ARIA will all now behave differently from AES. When writing or designing code for these, you want the interfaces and error codes to be the same, and differences are confusing. 

We deprecated `MBEDTLS_ERR_xxxx_INVALID_KEY_LENGTH` to save error codes, and because we don't want multiple error codes for what is essentially the same error - invalid parameters.

Semantically, we should have the *same* errors on each module for consistency. It should be uniform.

That doesn't mean we have to do it in this PR, (which I've promoted to 'ready to merge'), but we should either raise a bug or another PR to fix it.

In that PR we should probably make `MBEDTLS_ERR_AES_BAD_INPUT_DATA` the same value as `MBEDTLS_ERR_AES_INVALID_KEY_LENGTH` too.

@mpg / @hanno-arm - please say if you take a different view.

**Comment by sbutcher-arm** [comment]
> No, MBEDTLS_ERR_AES_BAD_INPUT_DATA was already present in development. It was introduced in #1178.

My mistake. I remember creating one, but that must have been an unnecessary fourth, which might have been removed by @k-stachowiak or me at an earlier point.

**Comment by mpg** [comment]
@sbutcher-arm Good point about checking whether the 3rd error code was already in development, and thanks @hanno-arm for confirming it was, I totally forgot to check and it would indeed have nullified my argument. Phew!

**Pull request merged**